### PR TITLE
rpc: Extract scriptPubKey on getreceivedbyaddress

### PR DESCRIPTION
**Pull request by konez2k**
Update rpc method `getreceivedbyaddress` to extract the correct scriptPubKey when shifted with opcodes (f.e. OP_CHECKSIG).
getreceivedbylabel is already implemented in the same way.

### CODE REVIEW

**Comment by MarcoFalke** [comment]
What is this refactoring or a bugfix?

**Comment by konez2k** [comment]
bugfix, the expected result from getreceivedbyaddress is different from getreceivedbylabel when the transaction scriptPubKey contains shifted opcodes

**Comment by MarcoFalke** [comment]
Would be nice to include a test that fails before the changes and passes after

**Comment by konez2k** [comment]
@MarcoFalke 
The issue is mainly with legacy addresses when sending a transaction using address pubkey

**Comment by sipa** [comment]
This will make P2PK outputs show up under the address for the corresponding P2PKH. Is that desirable?

**Comment by konez2k** [comment]
@sipa it is under `getreceivedbylabel`, shouldn't the output be equal between the 2 methods?

**Comment by sipa** [comment]
Ah, for consistency that does make sense.

**Comment by konez2k** [comment]
feature_block.py works for me (patch included/excluded), the error is probably due to a timeout on trevis side

File: test/functional/rpc_getreceivedbyaddress.py [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+                        "type": "pubkey",
+                        "addresses": [
+                            "mmo5AqgFVkaT8ehMN5wZ1Xec23h8Pu4T9q"
+                        ]
+                    }
+                }
+            ],
+            "hex": "02000000000100ca9a3b00000000232102352974bf0da1e6c4652888f014446572669cf7f9fd54d2c3b3d26ef60f15bd4cac00000000"
+        }
+        """
```
Comment (MarcoFalke) [inline test/functional/rpc_getreceivedbyaddress.py]: I think there is no need to copy paste the full output of the command

**Comment by luke-jr** [comment]
Concept NACK: p2pk aren't addresses, so this is actually introducing a bug.

Labels are [at least effectively] assigned for both p2pk and p2pkh, which is why there is a discrepancy.

**Comment by NicolasDorier** [comment]
Concept NACK for same reason than @luke-jr  and same reason as https://github.com/bitcoin/bitcoin/pull/16725 .

I lost so much time explaining why the genesis transaction output should not appear as an address in block explorers, only to get replied "but why bitcoin core does it?". This in general cause lot's of confusion to new developers and end up in bugs which happen in production when a malicious actor sends a payment as P2PK to the the same pubkey as another expected P2PKH.

**Comment by meshcollider** [comment]
@konez2k thanks for contributing! I think consistency is really valuable, but in this case as the others above have pointed out, it is probably consistency in the wrong direction. Unless anyone else is in support of this, I think its better to close it, but I hope that does not discourage you from contributing more in the future 😄

**Pull request closed without merging**